### PR TITLE
RA-113: Move CellRendererButton to VAS

### PR DESCRIPTION
**Pull request by vmartos**

### CODE REVIEW

File: VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+//
+using System;
+using Gdk;
+using GLib;
+using Gtk;
+using VAS.Core.Common;
+using VAS.Core.Interfaces.Drawing;
+using VAS.Drawing.Cairo;
+using Point = VAS.Core.Common.Point;
+using Image = VAS.Core.Common.Image;
```
Comment (aoller) [inline VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs]: Sort usings

File: VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs [hunk 2]
```diff
@@ -132,12 +132,15 @@ public override void GetSize (Widget widget, ref Rectangle cell_area, out int x_
...
 				tk.Begin ();
-				tk.FillColor = null;
+				tk.FillColor = VAS.Core.Common.Color.Green1;
+				if (flags.HasFlag (CellRendererState.Prelit)) {
+					tk.FillColor = VAS.Core.Common.Color.Green;
+				}
+
 				tk.LineWidth = 1;
-				tk.StrokeColor = App.Current.Style.PaletteBackgroundDark;
+				tk.StrokeColor = VAS.Core.Common.Color.Green;
```
Comment (aoller) [inline VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs]: Add a FIXME comment for future Style implementation.

File: VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+				tk.StrokeColor = App.Current.Style.PaletteText;
+				tk.FontAlignment = FontAlignment.Center;
+				tk.DrawText (pos, width, height, Text);
+				tk.End ();
+				tk.Context = null;
+			}
+		}
+	}
+
+	public class CellRendererButtonImage : CellRendererToggle
```
Comment (aoller) [inline VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs]: This should be in a new file

File: VAS.Core/Interfaces/GUI/IGUIToolkit.cs [hunk 4]
```diff
@@ -24,6 +24,7 @@
 using VAS.Core.Store;
 using VAS.Core.Store.Playlists;
 using Image = VAS.Core.Common.Image;
+using VAS.Core.Interfaces.MVVMC;
```
Comment (aoller) [inline VAS.Core/Interfaces/GUI/IGUIToolkit.cs]: Sort usings

File: VAS.Core/MVVMC/ViewLocator.cs [hunk 5]
```diff
@@ -59,5 +59,18 @@ public IView Retrieve (string name)
 			}
 			return (IView)Activator.CreateInstance (elements [name]);
 		}
+
+		/// <summary>
+		/// Gets a View with the specified name. With constructor Params
+		/// </summary>
+		/// <param name="name">Name.</param>
+		/// <param name="args">View Constructor Param Arguments.</param>
+		public IView Retrieve (string name, params object[] args)
```
Comment (aoller) [inline VAS.Core/MVVMC/ViewLocator.cs]: Unused method

Comment (vmartos) [inline VAS.Core/MVVMC/ViewLocator.cs]: Is used in RiftAnalyst in:
https://bitbucket.org/fluendo/riftanalyst/commits/6761faddc35fd678f18361fd11522bc35d12d340?at=master#LRiftAnalyst.UI.Gtk2/Ui/GUIToolkit.csT292

To create the MatchFinderView Panel with a constructor that needs a parameter.

**Comment by aoller** [comment]
👍

File: VAS.Core/Common/StateController.cs [hunk 6]
```diff
@@ -230,18 +230,18 @@ public void EmptyStateStack ()
...
-		void PushNavigationState (string transition, IScreenState state)
+		async Task PushNavigationState (string transition, IScreenState state)
 		{
 			navigationStateStack.Add (new Tuple<string, IScreenState> (transition, state));
-			App.Current.Navigation.LoadNavigationPanelAsync (state.Panel);
+			await App.Current.Navigation.LoadNavigationPanelAsync (state.Panel);
 		}
 
-		void PopNavigationState ()
+		async Task PopNavigationState ()
```
Comment (vguzmanp) [inline VAS.Core/Common/StateController.cs]: Add a comment here telling why this is `async Task` and not `async void`

Comment (ylatuya) [inline VAS.Core/Common/StateController.cs]: This does not need comments. Async functions are only allowed to return void if it is an event handler

File: VAS.Core/Interfaces/GUI/IGUIToolkit.cs [hunk 7]
```diff
@@ -126,6 +127,8 @@ public interface IGUIToolkit
 		Task<bool> CreateNewTemplate<T> (IList<T> availableTemplates, string defaultName,
 		                                 string countText, string emptyText,
 		                                 CreateEvent<T> evt) where T: ITemplate;
+
+		void CreateNewPopUpView (string viewAttribute, IViewModel viewModel, object parent);
```
Comment (ylatuya) [inline VAS.Core/Interfaces/GUI/IGUIToolkit.cs]: The name of the function should be PopUpView, which is the real action done.
The function is undocumented.
This change is unrelated with the commit.

File: VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+using Gtk;
+using VAS.Core.Common;
+using VAS.Core.Interfaces.Drawing;
+using VAS.Drawing.Cairo;
+using Image = VAS.Core.Common.Image;
+using Point = VAS.Core.Common.Point;
+
+namespace VAS.UI.Component
+{
+	public delegate void ClickedHandler (object sender, ClickedArgs args);
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs]: This should be in the handlers file where all delegates are defined

File: VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+				tk.StrokeColor = App.Current.Style.PaletteText;
+				tk.FontAlignment = FontAlignment.Center;
+				tk.DrawText (pos, width, height, Text);
+				tk.End ();
+				tk.Context = null;
+			}
+		}
+	}
+
+	public class ClickedArgs : SignalArgs
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/TreeView/CellRendererButton.cs]: Should be moved to a common file, since it's also used in ther classes.

File: VAS.Core/Common/StateController.cs [hunk 10]
```diff
@@ -78,22 +222,103 @@ public bool UnRegister (string transition)
 
 			return ok;
 		}
+		//Empties the stacks
```
Comment (ylatuya) [inline VAS.Core/Common/StateController.cs]: bad docs format

File: VAS.Core/Interfaces/GUI/INavigation.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+{
+	public interface INavigation
+	{
+		/// <summary>
+		/// Loads the navigation panel async. It just removes the last panel from the main window
+		/// and replaces it with the new IPanel provided as argument.
+		/// </summary>
+		/// <returns>The navigation panel async.</returns>
+		/// <param name="panel">Panel.</param>
+		Task<bool> LoadNavigationPanelAsync (IPanel panel);
```
Comment (ylatuya) [inline VAS.Core/Interfaces/GUI/INavigation.cs]: The Async suffix is redundant

File: VAS.Core/Interfaces/GUI/INavigation.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+		/// <summary>
+		/// Loads the modal panel async. It creates a new ExternalWindow
+		/// </summary>
+		/// <returns>The modal panel async.</returns>
+		/// <param name="panel">Panel.</param>
+		/// <param name="parent">Parent.</param>
+		Task LoadModalPanelAsync (IPanel panel, IPanel parent);
```
Comment (ylatuya) [inline VAS.Core/Interfaces/GUI/INavigation.cs]: The Async suffix is redundant

File: VAS.UI.Gtk2/UI/Dialog/BusyDialog.cs [hunk 13]
```diff
@@ -0,0 +1,108 @@
+﻿//
+//  Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/Dialog/BusyDialog.cs]: Keep the copyright headers from longomatch.

File: VAS.UI.Gtk2/UI/Dialog/CalendarDialog.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
+﻿//
+//  Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/Dialog/CalendarDialog.cs]: Keep the copyright headers from longomatch.

File: VAS.UI.Gtk2/UI/Dialog/ChooseOptionDialog.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
+﻿//
+//  Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/Dialog/ChooseOptionDialog.cs]: Keep the copyright headers from longomatch.

File: VAS.UI.Gtk2/Dialogs.cs [hunk 16]
```diff
@@ -0,0 +1,152 @@
+﻿//
+//   Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.UI.Gtk2/Dialogs.cs]: Keep same copyright headers as from the previous file

File: VAS.Core/Interfaces/GUI/IDialogs.cs [hunk 17]
```diff
@@ -0,0 +1,51 @@
+﻿//
+//   Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.Core/Interfaces/GUI/IDialogs.cs]: Keep copyright headers

File: VAS.UI.Gtk2/GUIToolkit.cs [hunk 18]
```diff
@@ -0,0 +1,140 @@
+﻿//
+//   Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.UI.Gtk2/GUIToolkit.cs]: Keep copyright headers

File: VAS.UI.Gtk2/GUIToolkit.cs [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+using VAS.Core.Store;
+using VAS.Core.Store.Playlists;
+using Image = VAS.Core.Common.Image;
+
+namespace VAS.UI
+{
+	/// <summary>
+	/// IGUIToolkit common implementation
+	/// </summary>
+	public abstract class GUIToolkit : IGUIToolkit
```
Comment (ylatuya) [inline VAS.UI.Gtk2/GUIToolkit.cs]: Should probably be named GUIToolkitBase intead

Comment (ylatuya) [inline VAS.UI.Gtk2/GUIToolkit.cs]: Instead of being abstract, why not just remove IGUIToolkit  and let those implementing the interface inherit from it ?

File: VAS.Core/Common/AsyncHelpers.cs [hunk 20]
```diff
@@ -0,0 +1,33 @@
+﻿//
+//   Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.Core/Common/AsyncHelpers.cs]: missing GPL license header

File: VAS.Core/Interfaces/GUI/INavigation.cs [hunk 21]
```diff
@@ -0,0 +1,34 @@
+﻿//  Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.Core/Interfaces/GUI/INavigation.cs]: GPL headers

File: VAS.UI.Gtk2/Dialogs.cs [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+			get {
+				return ((GUIToolkitBase)App.Current.GUIToolkit).GetParentWidget (null) as Gtk.Window;
+			}
+		}
+
+		#region IDialogs implementation
+
+		public void InfoMessage (string message, object parent = null)
+		{
+			MessagesHelpers.InfoMessage (((GUIToolkitBase)App.Current.GUIToolkit).GetParentWidget (parent), message);
```
Comment (ylatuya) [inline VAS.UI.Gtk2/Dialogs.cs]: (((GUIToolkitBase)App.Current.GUIToolkit).GetParentWidget is used several times, can't we make a helper function for it?

File: VAS.UI.Gtk2/GUIToolkitBase.cs [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+		protected Gtk.Window MainWindow {
+			get {
+				return mainWindow; 
+			}
+			set {
+				mainWindow = value;
+			}
+		}
+
+		Gtk.Window mainWindow;
```
Comment (ylatuya) [inline VAS.UI.Gtk2/GUIToolkitBase.cs]: Variables first

File: VAS.UI.Gtk2/GUIToolkitBase.cs [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+using Image = VAS.Core.Common.Image;
+
+namespace VAS.UI
+{
+	/// <summary>
+	/// IGUIToolkit common implementation
+	/// </summary>
+	public abstract class GUIToolkitBase : IGUIToolkit
+	{
+		protected Gtk.Window MainWindow {
```
Comment (ylatuya) [inline VAS.UI.Gtk2/GUIToolkitBase.cs]: Properties after contstructors

File: VAS.UI.Gtk2/UI/TreeView/CellRendererButtonImage.cs [hunk 25]
```diff
@@ -0,0 +1,77 @@
+﻿//  Copyright (C) 2016 Fluendo S.A.
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/TreeView/CellRendererButtonImage.cs]: GPL header

**Comment by ylatuya** [comment]
👍
